lib/src/async-compiler.ts

```diff
@@ -6,42 +6,180 @@ import {spawn} from 'child_process';
+ * differentiate and throw an error if the `AsyncCompiler` is constructed via
+ * `new AsyncCompiler`.
  */
-export class AsyncEmbeddedCompiler {
+const initFlag = Symbol();
+
+/**
+ * An asynchronous wrapper for the embedded Sass compiler
+ */
+export class AsyncCompiler {
   /** The underlying process that's being wrapped. */
   private readonly process = spawn(
     compilerCommand[0],
     [...compilerCommand.slice(1), '--embedded'],
     {windowsHide: true}
   );
 
+  /** Whether the underlying compiler has already exited. */
+  private disposed = false;
+
+  /** Reusable message transformer for all compilations.  */
+  private messageTransformer: MessageTransformer;
+
+  /** A list of pending compilations */
+  private compilations = new Set<Promise<CompileResult>>();
+
   /** The child process's exit event. */
-  readonly exit$ = new Promise<number | null>(resolve => {
+  private readonly exit$ = new Promise<number | null>(resolve => {
     this.process.on('exit', code => resolve(code));
   });
 
   /** The buffers emitted by the child process's stdout. */
-  readonly stdout$ = new Observable<Buffer>(observer => {
+  private readonly stdout$ = new Observable<Buffer>(observer => {
     this.process.stdout.on('data', buffer => observer.next(buffer));
   }).pipe(takeUntil(this.exit$));
 
   /** The buffers emitted by the child process's stderr. */
-  readonly stderr$ = new Observable<Buffer>(observer => {
+  private readonly stderr$ = new Observable<Buffer>(observer => {
     this.process.stderr.on('data', buffer => observer.next(buffer));
   }).pipe(takeUntil(this.exit$));
 
   /** Writes `buffer` to the child process's stdin. */
-  writeStdin(buffer: Buffer): void {
+  private writeStdin(buffer: Buffer): void {
     this.process.stdin.write(buffer);
   }
 
-  /** Kills the child process, cleaning up all associated Observables. */
-  close() {
+  /** Adds a compilation to the pending set and removes it when it's done. */
+  private addCompilation(compilation: Promise<CompileResult>): void {
+    this.compilations.add(compilation);
+    compilation
+      .catch(() => {})
+      .finally(() => this.compilations.delete(compilation));
+  }
+
+  /** Guards against using a disposed compiler. */
+  private throwIfDisposed(): void {
+    if (this.disposed) {
+      throw utils.compilerError('Async compiler has already been disposed.');
+    }
+  }
+
+  /**
+   * Sends a compile request to the child process and returns a Promise that
+   * resolves with the CompileResult. Rejects the promise if there were any
+   * protocol or compilation errors.
+   */
+  private async compileRequestAsync(
+    request: proto.InboundMessage_CompileRequest,
+    importers: ImporterRegistry<'async'>,
+    options?: OptionsWithLegacy<'async'> & {legacy?: boolean}
+  ): Promise<CompileResult> {
+    const functions = new FunctionRegistry(options?.functions);
+
+    const dispatcher = createDispatcher<'async'>(
+      this.compilations.size + 1,
```

Comment:
This risks a race condition on compilation id conflict:

1. Send CompileRequest with `id = 1`,
2. Send CompileRequest with `id = 2`
3. Wait for CompileRequest `id = 1` finishes, and then wait for `this.compilations.size` to drop back to 1.
4. Send another CompileRequest, and this one would also get `id = 2`, which conflicts with an existing request.

Here is a demo:

``` js
import * as sass from 'sass-embedded';

const delay = function (value, duration) {
  return new Promise((resolve, _) => setTimeout(() => resolve(value), duration));
}

const main = async function () {
  const asyncCompiler = await sass.initAsyncCompiler();

  const future1 = asyncCompiler.compileStringAsync('a { b: foo() }', {
    functions: {
      'foo()': (_) => delay(new sass.SassString('1'), 1000)
    }
  })

  const future2 = asyncCompiler.compileStringAsync('a { b: foo() }', {
    functions: {
      'foo()': (_) => delay(new sass.SassString('2'), 2000)
    }
  })

  console.log((await future1).css);

  // Because `this.compilations.delete(compilation)` runs asynchronously,
  // we need to wait a bit of time in order to reliably reproduce the issue. 
  await delay(null, 100);

  const future3 = asyncCompiler.compileStringAsync('a { b: 3 }')

  console.log((await future2).css);
  console.log((await future3).css);

  await asyncCompiler.dispose();
}

main()
```

Output:

```
a {
  b: "1";
}
Host caused params error: A CompileRequest with compilation ID 2 is already active.
```

Reply:
Something like this would work:

``` diff
diff --git a/lib/src/async-compiler.ts b/lib/src/async-compiler.ts
index 66611c7..f4f82ac 100644
--- a/lib/src/async-compiler.ts
+++ b/lib/src/async-compiler.ts
@@ -48,6 +48,9 @@ export class AsyncCompiler {
   /** Reusable message transformer for all compilations.  */
   private messageTransformer: MessageTransformer;
 
+  /* The next compilationId */
+  private id = 1;
+
   /** A list of pending compilations */
   private compilations = new Set<Promise<CompileResult>>();
 
@@ -76,7 +79,10 @@ export class AsyncCompiler {
     this.compilations.add(compilation);
     compilation
       .catch(() => {})
-      .finally(() => this.compilations.delete(compilation));
+      .finally(() => {
+        this.compilations.delete(compilation);
+        if (this.compilations.size === 0) this.id = 1;
+      });
   }
 
   /** Guards against using a disposed compiler. */
@@ -99,7 +105,7 @@ export class AsyncCompiler {
     const functions = new FunctionRegistry(options?.functions);
 
     const dispatcher = createDispatcher<'async'>(
-      this.compilations.size + 1,
+      this.id++,
       this.messageTransformer,
       {
         handleImportRequest: request => importers.import(request),

```

Reply:
Thanks for catching that. I implemented it in b03ef756129d882b1cba6bc3b1e5cd7457f79604

Reply:
I also refactored the async compiler to not require the extra `addCompilation` method. Instead, the promise and ID tracking are all handled in `compileRequestAsync`, keeping all related code in one place. See 86d5bcc29d1ff5456ea44735b2c4066429795932

Reply:
As @nex3 pointed out, let's just keep both implementation the same.